service/header/sync.go

```diff
@@ -20,147 +23,263 @@ type Syncer struct {
 	// is set to 0 once syncing is either finished or
 	// not currently in progress
 	inProgress uint64
+	// signals to start syncing
+	triggerSync chan struct{}
+	// pending keeps ranges of valid headers rcvd from the network awaiting to be appended to store
+	pending ranges
+
+	cancel context.CancelFunc
 }
 
 // NewSyncer creates a new instance of Syncer.
-func NewSyncer(exchange Exchange, store Store, trusted tmbytes.HexBytes) *Syncer {
+func NewSyncer(exchange Exchange, store Store, sub *P2PSubscriber, trusted tmbytes.HexBytes) *Syncer {
 	return &Syncer{
-		exchange:   exchange,
-		store:      store,
-		trusted:    trusted,
-		inProgress: 0, // syncing is not currently in progress
+		sub:         sub,
+		exchange:    exchange,
+		store:       store,
+		trusted:     trusted,
+		triggerSync: make(chan struct{}, 1), // should be buffered
 	}
 }
 
 // Start starts the syncing routine.
-func (s *Syncer) Start(context.Context) error {
-	go s.Sync(context.TODO()) // TODO @Wondertan: leaving this to you to implement in disconnection toleration PR.
+func (s *Syncer) Start(ctx context.Context) error {
+	if s.cancel != nil {
+		return fmt.Errorf("header: Syncer already started")
+	}
+
+	if s.sub != nil {
+		err := s.sub.AddValidator(s.validateMsg)
+		if err != nil {
+			return err
+		}
+	}
+
+	// TODO(@Wondertan): Ideally, this initialization should be part of Init process
+	err := s.initStore(ctx)
+	if err != nil {
+		log.Error(err)
```

Comment:
Returning error here breaks tests in `node_test.go` because trustedHash there is not provided. Even if we provide it, syncer will try to fetch the header by hash, but there is nowhere it can fetch so this always errors, so I making this more like an optional thing.

Reply:
Hmmm.... still seems weird though.

Reply:
Yes, this is why this initialization part should live elsewhere not cause such issues, but we can live with this for now

Reply:
And this is related to the comment above